### PR TITLE
Introduction of concepts for next-generation call resolving

### LINKED ISSUE #406: Refactor CallResolver to allow for more precise resolutions
The CallResolver is getting increasingly hard to maintain, since we need to distinguish between the call resolution for C++ and for Java. This will probably become worse as we want to resolve function calls of experimental languages. @konradweiss suggested to incorporate different CallResolvers based on the language with a strategy pattern, so that we can provide more precise call resolution. @konradweiss  also noted that it is probably not sufficient to check for the LanguageFrontend as we are doing currently, but instead check the language on a lower level (nodes?). The reason for this would be that there are ways to mix multiple languages, even in the same TUD. If we discuss changes in the architecture of the passes, maybe we should also take #310 into consideration.

### LINKED ISSUE #1497: Migrate regular function call resolving to new resolver

### LINKED ISSUE #1498: Provide an initial implementation and concepts for the next-generation resolving
We should probably start with the implementation of the necessary concepts and classes and fill in the necessary placeholders step by step.

### PR DESCRIPTION
**Pull request by oxisto**
This PR tries to implement the initial concepts described in #704 and also migrates the resolving of regular function calls to the new system.

Fixes #406
Fixes #1498 
Fixes #1497

### CODE REVIEW

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:715]: Doc at least for PROBLEMATIC (= a state that should never happen) and AMBIGUOUS (= can happen but there are multiple options e.g. due to overriding. What's the issue here that prevents us from saying "OK"?)

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:701]: It's not quite clear why we need the `candidateFunctions` or the `call` in the result. It's either the input or an intermediate step, so it's not super relevant? Some doc for all the lists, sets, maps would be useful. Why is `candidateFunctions` a list and not a set (same for `bestViable`)?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:64]: Maybe merge with `CastNeeded`?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:62]: Just `DirectMatch`?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:395]: I'd somehow prefer to set the `result.type` here instead of having the default `OK` set in the `ScopeManager`

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:390]: This should be fine if we have languages where we can overload or override functions/methods. Otherwise it would be PROBLEMATIC (e.g. in C), right? Can we use our LanguageTraits here to make this choice?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/FunctionDeclaration.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/FunctionDeclaration.kt:109]: Necessary/Ever used? Remove from this PR or implement it ;)

File: cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CLanguage.kt
Comment (KuechA) [inline cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CLanguage.kt:198]: Double-check spec if all numbers are on the same hierarchy level

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:525]: Is `log.error` appropriate here or is it only a warning?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (KuechA) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:537]:
```suggestion
                // declaration or a translation unit. Nothing else is allowed (for now). We can
```

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:525]: A warning is probably better.

File: cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CLanguage.kt
Comment (oxisto) [reply cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/CLanguage.kt:198]: It seems so, because if you have something like this:
```cpp
void f(int x) {}
void f(float x) {}

int main() {
   f(2.0); // 2.0 is a double
}
```

Then clang says its ambiguous, so it seems that double to float has the same ranking as double to int.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:525]: A would keep an error for "PROBLEMATIC" though

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:715]: Do you have a better naming idea for `Type`?

**Comment by KuechA** [comment]
Did you check if this PR fixes #1045?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:701]: I would keep it to give the language refinement method as much information as possible, if a language decides to implement something very very accurate. Also its nice to have from a debugging perspective if you look at a finished result and see what the original candidates were.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt]: I renamed it to `SuccessKind` for now

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt]: We probably need to differentiate between overloading and overriding. Even if you overload a function, there can be a point, in which we cannot differentiate between two function candidates, which are equally ranked and thus ambiguous. A compiler will return an error. I described this now in the doc.

I am not 100 % sure how we will deal with overridden functions, in this case they could still both match but the result is probably "ok". Not sure if we will only encounter this once I try to implement method resolution.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/FunctionDeclaration.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/declarations/FunctionDeclaration.kt:109]: Not used. can be removed

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:400]: I can try, the only issue is that we need to set the PROBLEMATIC already in resolveCall when we have a language that does not allow function overloading but more than one candidate function was returned. We are basically not allowed to "recover" from the problematic state, so that's why I handle it in resolveCall.

Open to suggestions :)

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt]: I agree that we have to differentiate between these two issues and that it's probably a future topic.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:705]: ok, so we keep them. Remaining points:
* Documentation of `candidateFunctions`, `viableFunctions`, `signatureResults`, `bestViable`
* `List` vs `Set`: There's no ordering and no duplicates, so a `Set` should work in all cases

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (KuechA) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:400]: Hmm but couldn't we currently recover from it if there were multiple direct matches (which would be very weird) or if there is no direct match but several equally-ranked matches? Then, we should probably stop the matching after l. 796 in ScopeManager and set the remaining fields there or at least, we should not enter this function (which we do at the moment)?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt]: Added more documentation and made all of them a set.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt]: > Hmm but couldn't we currently recover from it if there were multiple direct matches (which would be very weird) or if there is no direct match but several equally-ranked matches? 

Ah you are correct, in this case we would "upgrade" from problematic to ambiguous, which is not intended.

> Then, we should probably stop the matching after l. 796 in ScopeManager and set the remaining fields there or at least, we should not enter this function (which we do at the moment)?

I think it's probably easiest if we do not enter bestViableResolution in this case then. That's how I implemented it now.

**Comment by oxisto** [comment]
> Did you check if this PR fixes #1045?

I theory yes, in practice this still fails because of #194, which we probably should fix first then.

**Review by KuechA — APPROVED** [review]

**Comment by oxisto** [comment]
> > Did you check if this PR fixes #1045?
> 
> I theory yes, in practice this still fails because of #194 because the function definition and declaration are not connected and both end up as candidates and we therefore think this is a problematic result :(

After #1484, this now also fixes #1045, at least for the case where the stdlib is included. Otherwise we cannot know that `size_t` is an integer and it will not match.

**Comment by oxisto** [comment]
@KuechA I will wait until next week to merge this after @maximiliankaul  and @konradweiss are back for some more discussions. It is important to get the concepts correct this time.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (konradweiss) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:1078]: Maybe matchingFunctions would be a better name, signatureMatchingFunctions is a bit lengthy, but to me the difference between candidateFunctions and viable functions is not clear enough from the name alone.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (konradweiss) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:283]: Are there no casts where our type classes can differ? I intuitively think of wrappers for primitive types in java but I don't know if our modeling currently has an issue with the casting here.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (konradweiss) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:191]: I would like a small explanation here why we do need it to find better places for it in future refactoring.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (konradweiss) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:280]: Is this faster than using a filter like `list.filter {...}` ?

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (konradweiss) [inline cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:519]: I would prefere a comment detaching this from the next gen such that it is more understandable in the future and renaming the variable to useNonLegacyResolution or negating the bools and call it `useLegacyResolution`

**Review by konradweiss — CHANGES_REQUESTED** [review]
The functionality looks like what we want and I have only some minor remarks on documentation and naming.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt:1078]: I took the naming from the C++ overload resolution algorithm: https://en.cppreference.com/w/cpp/language/overload_resolution#Viable_functions

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt:283]: Unboxing of primitive types in Java is probably something that needs to be specified in the Java language

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:280]: Hmm it is probably a little bit faster since we are removing elements from the list that we need to check, so if we have a large list of function declarations, using the mutable iterator should be faster since we are removing those declarations step by step. Using the filter on the original list would mean that we need to loop through the original list items even though we know that the element was already removed.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt]: I already moved it to the "better place"

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt
Comment (oxisto) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt]: Good point, it was a bit too much kotlin magic. I restructed the function for better readability.

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt
Comment (konradweiss) [reply cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/SymbolResolver.kt:283]: Ah there is the check if its definition is already in the list so we cannot merge the entire conditional removing into the filter or predicate of the scope resolution above. At least not trivially

**Review by konradweiss — APPROVED** [review]

**Pull request merged**
